site/content/blog/2020-05-20-hands-on-iouring-go.md

```diff
@@ -0,0 +1,327 @@
+
+All that is changing slowly, because now we have a brand new interface to perform I/O with the kernel - io_uring. There's a lot of buzz happening around it. And rightly so, because it gives us an entirely new model to interact with the kernel. Let's dive into it and try to understand what it is and how does it solve the problem. And then we will construct a small demo application with Go to play with it.
+
+### Background
+
+Let's take a step back and think of how usual system calls work. We make a syscall, our application in user layer calls into the kernel, it makes a copy of the data in kernel space. After kernel is done with the execution, it copies the result back to user-space buffers. And then it returns. All this while, the syscall remains blocked.
+
+![image](/blog/2020-05-20-hands-on-iouring-go/syscall.png)
+
+Right off the bat, we can see a lot of bottlenecks - there's lots of copying going around, and there's blocking. Go handles this problem by bringing another layer between the application and the kernel - the runtime. It uses a virtual entity (commonly referred to as P) which contains a queue of goroutines to run, which is then mapped to OS threads. This level of indirection allows it to do some interesting optimizations. Whenever we make a blocking syscall, the runtime is aware of it, and it detaches the thread from the P executing the goroutine, and acquires a new thread to execute other goroutines. This is known as a hand-off. And when the syscall returns, the runtime tries to re-attach it to a P. If it cannot get a free P, it just pushes the goroutine to a queue to be executed later, and stores the thread in a pool. This is how Go gives the appearance of "non-blocking"-ness when your code enters a system call.
+
+This is great, but it still does not solve the main problem - which is that copies still happen, and the actual syscall still blocks.
+
+Let's think of the first problem at hand - copies. How do we prevent copies happening from user-space to kernel-space ? Well, obviously we need some sort of shared memory. Ok, that can be done using the `mmap` system call which can map a chunk of memory that is shared between the user and kernel. That takes care of copying, but what about synchronization ? Even if we don't copy, we need some way to synchronize data access between us and the kernel. Otherwise we fall into the same problem - because the application would need to make a syscall again to perform the locking.
```

Comment:
```suggestion
Let's think of the first problem at hand - copies. How do we prevent copies happening from user-space to kernel-space? Well, obviously we need some sort of shared memory. Ok, that can be done using the `mmap` system call which can map a chunk of memory that is shared between the user and kernel. 

That takes care of copying, but what about synchronization? Even if we don't copy, we need some way to synchronize data access between us and the kernel. Otherwise we encounter the same problem - because the application would need to make a syscall again to perform the blocking.
```

Reply:
Not sure if my suggestion for "locking" to "blocking" is correct in this context.

Reply:
Oops, I totally missed that. "locking" is the right word here :) The application needs to make a syscall to lock the data it's accessing.

Reply:
Awesome, thanks for confirming. :)